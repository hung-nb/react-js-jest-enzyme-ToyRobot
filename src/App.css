/* Flex area */
.App {
  display: flex;
  flex-direction: row;
  text-align: center;
  width: 600px;
  background: #eee;
  border: 1px solid #ddd;
  padding: 10px;
  margin: 20px auto;
}

.mainArea {
  flex: 1;
  margin: 15px;
  text-align: center;
}

.logArea {
  flex: 1;
  margin: 15px;
  text-align: center;
}

/* Common */
.input {
  display: block;
  width: 100%;
  margin: 5px 0;
  padding: 5px 0;
  box-sizing: border-box;
  font-size: 15px;
  text-align: center;
}

.textArea {
  display: block;
  width: 100%;
  height: 300px;
  margin: 5px 0;
  padding: 5px 0;
  box-sizing: border-box;
  font-size: 15px;
  text-align: center;
}
