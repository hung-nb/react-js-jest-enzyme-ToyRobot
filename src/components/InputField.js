import React from 'react'

const InputField = (props) => {
    return (
        <div>
            <div>Input Commands</div>
            <input
                className="input"
                type="text"
                placeholder = "Enter a command:"
                onChange = {props.onChange}
                onKeyPress={props.onKeyPress}
                value={props.value}>
            </input>
        </div>
    )
}

export { InputField };