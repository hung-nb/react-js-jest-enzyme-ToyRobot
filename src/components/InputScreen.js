import React from 'react'

const InputScreen = (props) => {
    return (
        <div>
            <div>Input Commands</div>
            <input 
                type="text"
                placeholder = "Enter a command:"
                onChange = {props.onChange}
                onKeyPress={props.onKeyPress}
                value={props.value}>
            </input>
        </div>
    )
}

export { InputScreen };