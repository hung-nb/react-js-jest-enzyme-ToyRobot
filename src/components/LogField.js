import React from 'react'

const LogField = (props) => {
    return (
        <div>
            <div>Input History</div>
            <textarea
                readOnly
                className='textArea'
                type="text"
                value = {props.value}>
            </textarea>
        </div>
    )
}

export { LogField };