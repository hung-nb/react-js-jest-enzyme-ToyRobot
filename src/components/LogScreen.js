import React from 'react'

const LogScreen = (props) => {
    return (
        <div>
            <div>Command History</div>
            <textarea
                type="text"
                value = {props.value}>
            </textarea>
        </div>
    )
}

export { LogScreen };