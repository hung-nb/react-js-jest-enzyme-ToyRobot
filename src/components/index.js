export * from './InputField';
export * from './ReportField';
export * from './LogField';