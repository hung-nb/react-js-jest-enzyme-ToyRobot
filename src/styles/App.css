/* Flex area */
.App {
  display: flex;
  flex-direction: row;
  align-items: center;
  width: 600px;
  background:#ddd;
  border: 1px solid #ddd;
  padding: 10px;
  margin: 20px auto;
}

.mainArea {
  flex: 1;
  margin: 15px;
  text-align: center;
}

.logArea {
  flex: 1;
  margin: 15px;
  text-align: center;
}

/* Common */
.input {
  display: block;
  width: 100%;
  margin: 5px 0;
  padding: 5px 0;
  box-sizing: border-box;
  font-size: 15px;
  text-align: center;
  font-weight: bold;
}

.textArea {
  display: block;
  width: 100%;
  height: 300px;
  margin: 5px 0;
  padding: 5px 0;
  box-sizing: border-box;
  font-size: 15px;
  text-align: center;
}
