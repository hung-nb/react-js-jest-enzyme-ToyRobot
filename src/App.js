import React, { Component } from 'react';
import { InputField, ReportField, LogField } from './components';
import {
  validatePlaceCommand,
  validatePosition,
  validateMoveCommand,
  validateLeftCommand,
  validateRightCommand,
  validateReportCommand,
} from './business/validateCommand-functions';
import {
  getNewPositionByMove,
  getNewPositionByLeft,
  getNewPositionByRight,
} from './business/getPosition-functions';

class App extends Component {

  //////////////////////
  // STATE CONTROLLER
  constructor(props) {        
		super(props);		
    this.state = {
      currentPosition: {
				X: -1,
				Y: -1,
				F: ""
		  },
      command: '',
      report: '',
      log: '',
    }
  }

  ///////////////////////////
  // USER ACTION CONTROLLER
  _handleInputChange(event) {
    this.setState({command: event.target.value})
  }
	
	_handleKeyPress(event) {

		if (event.key !== 'Enter') {
      return;
    }

    // LAYOUT
    let newCommand = this.state.command;
    this._clearCommand();

    // LOGIC
    let newPosition = this.state.currentPosition;
    if (newPosition.X === -1) {

      // PLACE command is needed
      newPosition = validatePlaceCommand(newCommand);
      if (newPosition === undefined) {
        return;
      }
      
      // PLACE
      this.setState({ currentPosition: newPosition });

    } else {

      if (validateReportCommand(newCommand)) {
        // REPORT
        let reportTxt = 'Output: ' 
          + newPosition.X + ','
          + newPosition.Y + ','
          + newPosition.F;
        this.setState({ report: reportTxt})
        return;
      }

      if (validatePlaceCommand(newCommand) !== undefined) {
        // PLACE
        newPosition = validatePlaceCommand(newCommand);
      }
      else if (validateMoveCommand(newCommand)) {
        // MOVE
        newPosition = getNewPositionByMove(newPosition);
      }
      else if (validateLeftCommand(newCommand)) {
        // LEFT
        newPosition = getNewPositionByLeft(newPosition);
      }
      else if (validateRightCommand(newCommand)) {
        // RIGHT
        newPosition = getNewPositionByRight(newPosition);
      }

      // Update current position
      if (validatePosition(newPosition))
        this.setState({ currentPosition: newPosition });
    }
  }

  _clearCommand() {
    let logTxt = this.state.log + '\n' + this.state.command;
    this.setState({ command: '', report: '' });
    this.setState({ log: logTxt });
  }
  
  ////////////////////
  // VIEW CONTROLLER
  render() {
    return (
      <div className="App">
        <div className='mainArea'>
          <InputField
            id="id_input"
            onChange={this._handleInputChange.bind(this)}
            onKeyPress={this._handleKeyPress.bind(this)}
            value={this.state.command}/>
          <br />
          <br />
          <ReportField
            value={this.state.report}/>
          <br />
          <br />
        </div>
        <div className='logArea'>
          <LogField
            value={this.state.log}/>
        </div>
      </div>
    );
  }
}

export default App;
