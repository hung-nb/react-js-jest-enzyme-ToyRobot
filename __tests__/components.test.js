import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { InputField, ReportField, LogField } from '../src/components/';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App from '../src/App';
configure({ adapter: new Adapter() });

/************ App */
test('State should be set correctly when command is inputted, case 1', () => {
    let state = {
        currentPosition: {
                  X: -1,
                  Y: -1,
                  F: ""
            },
        command: '',
        report: '',
        log: '',
    };
    const wrapper = shallow(<App {...state} />);
    const input = wrapper.find('#id_input');
    input.simulate('change', {target: {value: 'PLACE 0,0,NORTH'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'REPORT'}});
    input.simulate('keypress', {key: 'Enter'});
    expect(wrapper.state('report')).toEqual('Output: 0,1,NORTH');
});

test('State should be set correctly when command is inputted, case 2', () => {
    let state = {
        currentPosition: {
                  X: -1,
                  Y: -1,
                  F: ""
            },
        command: '',
        report: '',
        log: '',
    };
    const wrapper = shallow(<App {...state} />);
    const input = wrapper.find('#id_input');
    input.simulate('change', {target: {value: 'PLACE 0,0,NORTH'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'LEFT'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'REPORT'}});
    input.simulate('keypress', {key: 'Enter'});
    expect(wrapper.state('report')).toEqual('Output: 0,0,WEST');
});

test('State should be set correctly when command is inputted, case 3', () => {
    let state = {
        currentPosition: {
                  X: -1,
                  Y: -1,
                  F: ""
            },
        command: '',
        report: '',
        log: '',
    };
    const wrapper = shallow(<App {...state} />);
    const input = wrapper.find('#id_input');
    input.simulate('change', {target: {value: 'PLACE 1,2,EAST'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'LEFT'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'REPORT'}});
    input.simulate('keypress', {key: 'Enter'});
    expect(wrapper.state('report')).toEqual('Output: 3,3,NORTH');
});

test('State should be set correctly when command is inputted, case 4', () => {
    let state = {
        currentPosition: {
                  X: -1,
                  Y: -1,
                  F: ""
            },
        command: '',
        report: '',
        log: '',
    };
    const wrapper = shallow(<App {...state} />);
    const input = wrapper.find('#id_input');
    input.simulate('change', {target: {value: 'PLACE 1,2,EAST'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'LEFT'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'REPORT'}});
    input.simulate('keypress', {key: 'Enter'});
    expect(wrapper.state('report')).toEqual('Output: 3,3,NORTH');
});

test('State should be set correctly when command is inputted, case 5', () => {
    let state = {
        currentPosition: {
                  X: -1,
                  Y: -1,
                  F: ""
            },
        command: '',
        report: '',
        log: '',
    };
    const wrapper = shallow(<App {...state} />);
    const input = wrapper.find('#id_input');
    input.simulate('change', {target: {value: 'PLACE 5,5,NORTH'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'LEFT'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'REPORT'}});
    input.simulate('keypress', {key: 'Enter'});
    expect(wrapper.state('report')).toEqual('Output: 4,5,WEST');
});

test('State should be set correctly when command is inputted, case 6', () => {
    let state = {
        currentPosition: {
                  X: -1,
                  Y: -1,
                  F: ""
            },
        command: '',
        report: '',
        log: '',
    };
    const wrapper = shallow(<App {...state} />);
    const input = wrapper.find('#id_input');
    input.simulate('change', {target: {value: 'PLACE 0,0,SOUTH'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'LEFT'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'REPORT'}});
    input.simulate('keypress', {key: 'Enter'});
    expect(wrapper.state('report')).toEqual('Output: 1,0,EAST');
});

test('State should be set correctly when command is inputted, case 7', () => {
    let state = {
        currentPosition: {
                  X: -1,
                  Y: -1,
                  F: ""
            },
        command: '',
        report: '',
        log: '',
    };
    const wrapper = shallow(<App {...state} />);
    const input = wrapper.find('#id_input');
    input.simulate('change', {target: {value: 'PLACE 5,5,EAST'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'RIGHT'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'REPORT'}});
    input.simulate('keypress', {key: 'Enter'});
    expect(wrapper.state('report')).toEqual('Output: 5,4,SOUTH');
});

test('State should be set correctly when command is inputted, case 8', () => {
    let state = {
        currentPosition: {
                  X: -1,
                  Y: -1,
                  F: ""
            },
        command: '',
        report: '',
        log: '',
    };
    const wrapper = shallow(<App {...state} />);
    const input = wrapper.find('#id_input');
    input.simulate('change', {target: {value: 'PLACE 0,5,WEST'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'RIGHT'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'REPORT'}});
    input.simulate('keypress', {key: 'Enter'});
    expect(wrapper.state('report')).toEqual('Output: 0,5,NORTH');
});

test('State should be set correctly when command is inputted, case 9', () => {
    let state = {
        currentPosition: {
                  X: -1,
                  Y: -1,
                  F: ""
            },
        command: '',
        report: '',
        log: '',
    };
    const wrapper = shallow(<App {...state} />);
    const input = wrapper.find('#id_input');
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'RIGHT'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'REPORT'}});
    input.simulate('keypress', {key: 'Enter'});
    expect(wrapper.state('report')).toEqual('');
});

test('State should be set correctly when command is inputted, case 10', () => {
    let state = {
        currentPosition: {
                  X: -1,
                  Y: -1,
                  F: ""
            },
        command: '',
        report: '',
        log: '',
    };
    const wrapper = shallow(<App {...state} />);
    const input = wrapper.find('#id_input');
    input.simulate('change', {target: {value: 'place 0,1,NORTH'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'RIGHT'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'REPORT'}});
    input.simulate('keypress', {key: 'Enter'});
    expect(wrapper.state('report')).toEqual('');
});

test('State should be set correctly when command is inputted, case 10', () => {
    let state = {
        currentPosition: {
                  X: -1,
                  Y: -1,
                  F: ""
            },
        command: '',
        report: '',
        log: '',
    };
    const wrapper = shallow(<App {...state} />);
    const input = wrapper.find('#id_input');
    input.simulate('change', {target: {value: 'PLACE 0,1,NORTH'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVe'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE '}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'RIGHT'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'REPORT'}});
    input.simulate('keypress', {key: 'Enter'});
    expect(wrapper.state('report')).toEqual('Output: 1,1,EAST');
});

test('State should be set correctly when command is inputted, case 10', () => {
    let state = {
        currentPosition: {
                  X: -1,
                  Y: -1,
                  F: ""
            },
        command: '',
        report: '',
        log: '',
    };
    const wrapper = shallow(<App {...state} />);
    const input = wrapper.find('#id_input');
    input.simulate('change', {target: {value: 'PLACE 0,1,NORTH'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'RIGHT'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'MOVE'}});
    input.simulate('keypress', {key: 'Enter'});
    input.simulate('change', {target: {value: 'REPORT'}});
    input.simulate('keypress', {key: 'Enter'});
    expect(wrapper.state('report')).toEqual('Output: 1,2,EAST');
});

/*********** InputField */
test('InputField should render correctly', () => {
    const value = 'PLACE 1,2,SOUTH';
    const result = shallow(
        <InputField value={value} />);
    expect(shallowToJson(result)).toMatchSnapshot();
});

/*********** ReportField */
test('ReportField should render correctly', () => {
    const value = 'Output: 1,2,SOUTH';
    const result = shallow(
        <ReportField value={value} />);
    expect(shallowToJson(result)).toMatchSnapshot();
});

/*********** LogField */
test('LogField should render correctly', () => {
    const value = 'PLACE 1,2,SOUTH';
    const result = shallow(
        <LogField value={value} />);
    expect(shallowToJson(result)).toMatchSnapshot();
});