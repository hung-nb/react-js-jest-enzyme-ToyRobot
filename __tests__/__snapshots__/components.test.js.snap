// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField should render correctly 1`] = `
<div>
  <div>
    Input
  </div>
  <input
    className="input"
    placeholder="Enter a command:"
    type="text"
    value="PLACE 1,2,SOUTH"
  />
</div>
`;

exports[`LogField should render correctly 1`] = `
<div>
  <div>
    Input History
  </div>
  <textarea
    className="textArea"
    readOnly={true}
    type="text"
    value="PLACE 1,2,SOUTH"
  />
</div>
`;

exports[`ReportField should render correctly 1`] = `
<div>
  <div>
    Output
  </div>
  <input
    className="input"
    readOnly={true}
    type="text"
    value="Output: 1,2,SOUTH"
  />
</div>
`;
